File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        int count = 0;
        int minVal;

        kthSmallestHelper(root, k, count, minVal);

        return minVal;
    }

    void kthSmallestHelper(TreeNode* root, int k, int &count, int &minVal) {
        if (root == nullptr) {
            return;
        }

        kthSmallestHelper(root->left, k, count, minVal);

        if (count >= k) return;

        count += 1;
        if (count == k) {
            minVal = root->val;
            return;
        }

        kthSmallestHelper(root->right, k, count, minVal);
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def dfs(node):
            if not node:
                return [0, 0]

            leftDiameter, leftHeight = dfs(node.left)
            rightDiameter, rightHeight = dfs(node.right)

            diameter = max(leftDiameter, rightDiameter, leftHeight + rightHeight)
            height = 1 + max(leftHeight, rightHeight)

            return [diameter, height]

        return dfs(root)[0]

        